dashboard/app/controllers/student_code_sample_controller.rb

```diff
@@ -3,49 +3,112 @@ class StudentCodeSampleController < ApplicationController
-          body = s3.get_object(bucket: bucket, key: s3_filename)[:body].read
-          student_code = JSON.parse(body)['source'] if body
-          code_samples << {level_id: level_id, script_id: script_id, user_id: student_id, project_id: token, student_code: student_code}
+        student_code = get_student_code(student_id, level_id, unit_id)
+        if student_code[:student_code]
+          code_samples << {level_id: level_id, unit_id: unit_id, user_id: student_id, project_id: student_code[:project_id], student_code: student_code[:student_code]}
+        end
+        have_enough_samples = code_samples.length >= num_samples
+      end
+    end
+    render json: code_samples
+  end
+
+  def fetch_student_code_samples_with_evaluations(level_id, unit_id, num_samples)
+    evaluations = UserLevelEvaluation.where(level_id: level_id, script_id: unit_id)
+    if evaluations.empty?
+      return render status: :not_found, json: "There are no evaluations for the level with id #{level_id} in unit with id #{unit_id}"
+    end
+    code_samples = []
+    have_enough_samples = false
+    evaluations.each do |evaluation|
+      unless have_enough_samples
+        student_code = get_student_code(evaluation.user_id, level_id, unit_id, evaluation.code_version)
+        if student_code[:student_code]
+          code_samples << {
+            level_id: level_id,
+            unit_id: unit_id,
+            user_id: evaluation.user_id,
+            project_id: student_code[:project_id],
+            code_version: student_code[:code_version],
+            student_code: student_code[:student_code],
+            ai_evaluation: evaluation.ai_evaluation,
+            ai_reasoning: evaluation.ai_reasoning,
+            evaluation_criteria: evaluation.evaluation_criteria
+          }
         end
         have_enough_samples = code_samples.length >= num_samples
       end
     end
     render json: code_samples
   end
+
+  def get_student_code(user_id, level_id, unit_id, code_version = nil)
+    s3 = Aws::S3::Client.new
+    bucket = CDO.sources_s3_bucket
+    base_dir = CDO.sources_s3_directory
+
+    storage_id = storage_id_for_user_id(user_id)
+    channel_token = ChannelToken.where(storage_id: storage_id, level_id: level_id, script_id: unit_id).last
+    user_level = UserLevel.where(user_id: user_id, level_id: level_id, script_id: unit_id).last
+    if user_level && channel_token
+      storage_app_id = channel_token.storage_app_id
+      token = storage_encrypt_channel_id(storage_id, storage_app_id)
+      _owner_storage_id, project_id = storage_decrypt_channel_id(token)
+      s3_filename = "#{base_dir}/#{storage_id}/#{project_id}/main.json"
+      s3_args = {bucket: bucket, key: s3_filename}
+      s3_args[:version_id] = code_version if code_version
+      body = s3.get_object(s3_args)[:body].read
+      student_code = JSON.parse(body)['source'] if body
+    end
+    {
+      project_id: token,
```

Comment:
admitting here our terminology is difficult...
* `storage_app_id` refers specifically to the unencrypted id for a project, but is kind of clunky as a name
* `channel_id` refers to the encrypted id. also not a great name, but is more or less the definitive name for this in our system
* `project_id` is more commonly used to describe the unencrypted project id, though you'll probably find some places it refers to the encrypted value

if it is not too difficult or confusing to change, my recommendation would be to call this `channel_id` instead of `project_id`. 

please consider this optional though :-)

Reply:
ah, I think I see what you mean. `project_id` refers to the id on the Projects table in our database, whereas the `channel_id` is the unique identifier for where that project is stored in s3?

Reply:
I see what you mean. Thanks for the clarification! I updated the code that's getting the student code to better match this terminology, but kept `project_id` as the header for the csv column given the audience. I think for a contractor, "project id" more clearly conveys that the value in that column is the thing that uniquely identifies a project and can be used to generate a project view url.

Reply:
yes, if a person will see this, go with what will make most sense for them!